Tests/PureeTests/Output/BufferedOutputTests.swift

```diff
@@ -136,10 +142,84 @@ class BufferedOutputTests: XCTestCase {
         XCTAssertEqual(output.calledWriteCount, 4)
     }
 
+    func testParallelWrite() {
+        output.configuration.logEntryCountLimit = 3
+        output.configuration.retryLimit = 3
+
+        XCTAssertEqual(logStore.logs(for: "pv_TestingBufferedOutput").count, 0)
+        XCTAssertEqual(output.calledWriteCount, 0)
+
+        var writeCallbackCalledCount = 0
+        output.writeCallback = {
+            writeCallbackCalledCount += 1
+        }
+
+        let semaphore = DispatchSemaphore(value: 0)
+        let testIndices = 0..<200
+
+        for _ in testIndices {
+            DispatchQueue.global(qos: .background).async {
+                self.output.emit(log: self.makeLog())
+                semaphore.signal()
+            }
+        }
+
+        for _ in testIndices {
+            semaphore.wait()
+        }
+        output.resume()
+
+        let expectedWriteCount = 67
+        XCTAssertEqual(output.calledWriteCount, expectedWriteCount)
+        XCTAssertEqual(writeCallbackCalledCount, expectedWriteCount)
+    }
+
     override func tearDown() {
         super.tearDown()
 
         output.suspend()
         logStore.flush()
     }
 }
+
+class TestingBufferedOutputAsync: TestingBufferedOutput {
+    private var expectations: [XCTestExpectation] = []
+    var testCase: XCTestCase!
+
+    override var storageGroup: String {
+        return "pv_TestingBufferedOutput"
+    }
+
+    override func write(_ chunk: BufferedOutput.Chunk, completion: @escaping (Bool) -> Void) {
+        calledWriteCount += 1
+        let expectation = XCTestExpectation(description: "async writing")
```

Comment:
We cannot call like `testCase.expectation()` because this is not in the Main Thread.